File: scripts/file/export.js
/**
 * Export manager for various formats
 */

import { APIClient } from '../utils/api.js';

export class ExportManager {
    async export(format, content, filename) {
        switch (format) {
            case 'md':
                return this.exportMarkdown(content, filename);
            case 'html':
                return this.exportHTML(content, filename);
            case 'pdf':
                return this.exportPDF(content, filename);
            case 'docx':
                return this.exportDOCX(content, filename);
            default:
                throw new Error(`Unknown export format: ${format}`);
        }
    }

    exportMarkdown(content, filename) {
        const blob = new Blob([content], { type: 'text/markdown' });
        this.downloadBlob(blob, `${filename}.md`);
    }

    async exportHTML(content, filename) {
        try {
            const response = await APIClient.post('/export/html', {
                content,
                options: {
                    standalone: true,
                    toc: false
                }
            });

            if (!response.success) {
                throw new Error(response.error || 'HTML export failed');
            }

            const blob = new Blob([response.html], { type: 'text/html' });
            this.downloadBlob(blob, `${filename}.html`);

        } catch (error) {
            console.error('HTML export error:', error);
            throw error;
        }
    }

    async exportPDF(content, filename) {
        try {
            const response = await fetch('http://localhost:5000/api/export/pdf', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include',
                body: JSON.stringify({
                    content,
                    filename: `${filename}.pdf`,
                    options: {
                        paper_size: 'A4',
                        margin: '1in',
                        toc: false
                    }
                })
            });

            if (!response.ok) {
                throw new Error('PDF export failed');
            }

            const blob = await response.blob();
            this.downloadBlob(blob, `${filename}.pdf`);

        } catch (error) {
            console.error('PDF export error:', error);
            throw error;
        }
    }

    async exportDOCX(content, filename) {
        try {
            const response = await fetch('http://localhost:5000/api/export/docx', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include',
                body: JSON.stringify({
                    content,
                    filename: `${filename}.docx`,
                    options: {
                        toc: false
                    }
                })
            });

            if (!response.ok) {
                throw new Error('DOCX export failed');
            }

            const blob = await response.blob();
            this.downloadBlob(blob, `${filename}.docx`);

        } catch (error) {
            console.error('DOCX export error:', error);
            throw error;
        }
    }

    downloadBlob(blob, filename) {
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = filename;
        a.click();
        URL.revokeObjectURL(url);
    }
}

File: backend/routes/export.py
"""Document export routes."""
from flask import Blueprint, request, jsonify, send_file
import io

try:
    from backend.services.export_service import ExportService
    from backend.config import Config
except ImportError:
    from services.export_service import ExportService
    from config import Config

export_bp = Blueprint('export', __name__, url_prefix='/api/export')


def get_export_service() -> ExportService:
    """Get export service instance."""
    return ExportService(pandoc_path=Config.PANDOC_PATH)


@export_bp.route('/html', methods=['POST'])
def export_html():
    """Export markdown to HTML.

    Request JSON:
        {
            "content": "markdown text",
            "options": {
                "standalone": true,
                "css": "styles.css",
                "toc": false
            }
        }

    Returns:
        {
            "success": true,
            "html": "..."
        }
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({'error': 'No JSON data provided'}), 400

        content = data.get('content')
        options = data.get('options', {})

        if not content:
            return jsonify({'error': 'content is required'}), 400

        service = get_export_service()
        html = service.to_html(content, options)

        return jsonify({
            'success': True,
            'html': html
        })

    except RuntimeError as e:
        return jsonify({'error': str(e)}), 500
    except Exception as e:
        return jsonify({'error': f'Unexpected error: {str(e)}'}), 500


@export_bp.route('/pdf', methods=['POST'])
def export_pdf():
    """Export markdown to PDF.

    Request JSON:
        {
            "content": "markdown text",
            "filename": "document.pdf",
            "options": {
                "paper_size": "A4",
                "margin": "1in",
                "toc": false
            }
        }

    Returns:
        PDF file download
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({'error': 'No JSON data provided'}), 400

        content = data.get('content')
        filename = data.get('filename', 'document.pdf')
        options = data.get('options', {})

        if not content:
            return jsonify({'error': 'content is required'}), 400

        service = get_export_service()
        pdf_bytes = service.to_pdf(content, options)

        return send_file(
            io.BytesIO(pdf_bytes),
            mimetype='application/pdf',
            as_attachment=True,
            download_name=filename
        )

    except RuntimeError as e:
        return jsonify({'error': str(e)}), 500
    except Exception as e:
        return jsonify({'error': f'Unexpected error: {str(e)}'}), 500


@export_bp.route('/docx', methods=['POST'])
def export_docx():
    """Export markdown to DOCX.

    Request JSON:
        {
            "content": "markdown text",
            "filename": "document.docx",
            "options": {
                "toc": false
            }
        }

    Returns:
        DOCX file download
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({'error': 'No JSON data provided'}), 400

        content = data.get('content')
        filename = data.get('filename', 'document.docx')
        options = data.get('options', {})

        if not content:
            return jsonify({'error': 'content is required'}), 400

        service = get_export_service()
        docx_bytes = service.to_docx(content, options)

        return send_file(
            io.BytesIO(docx_bytes),
            mimetype='application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            as_attachment=True,
            download_name=filename
        )

    except RuntimeError as e:
        return jsonify({'error': str(e)}), 500
    except Exception as e:
        return jsonify({'error': f'Unexpected error: {str(e)}'}), 500
